h5py/_hl/dataset.py

```diff
@@ -198,47 +199,77 @@ def open_dset(parent, name, dapl=None, efile_prefix=None, virtual_prefix=None,
     return dset_id
 
 
-class AstypeWrapper:
-    """Wrapper to convert data on reading from a dataset.
-    """
-    def __init__(self, dset, dtype):
-        self._dset = dset
-        self._dtype = numpy.dtype(dtype)
 
-    def __getitem__(self, args):
-        return self._dset.__getitem__(args, new_dtype=self._dtype)
+class AbstractView(ABC):
+    _dset: "Dataset"
 
-    def __len__(self):
-        """ Get the length of the underlying dataset
+    def __init__(self, dset):
+        self._dset = dset
 
-        >>> length = len(dataset.astype('f8'))
-        """
+    def __len__(self):
         return len(self._dset)
 
-    def __array__(self, dtype=None, copy=True):
+    @property
+    def dtype(self):
+        return self._dset.dtype
+
+    @property
+    def ndim(self):
+        return self._dset.ndim
+
+    @property
+    def shape(self):
+        return self._dset.shape
+
+    @property
+    def size(self):
+        return self._dset.size
+
+    @abstractmethod
+    def __getitem__(self, idx):
+        ...  # pragma: nocover
+
+    def __array__(self, dtype=None, copy=None):
         if copy is False:
             raise ValueError(
-                f"AstypeWrapper.__array__ received {copy=} "
-                f"but memory allocation cannot be avoided on read"
+                f"{self.__class__.__name__}.__array__ received {copy=} "
+                "but memory allocation cannot be avoided on read"
             )
 
-        data = self[:]
+        data = self[()]
+        if not self.ndim:
+            return numpy.asarray(data, dtype=dtype or self.dtype)
         if dtype is not None:
             return data.astype(dtype, copy=False)
         return data
 
+class AstypeView(AbstractView):
```

Comment:
While we're renaming this anyway, we can make the naming pattern a bit more consistent (matching AsStrView).

```suggestion
class AsTypeView(AbstractView):
```